### ISSUE TITLE
Help from Rahul Bhalley #2

### ISSUE THREAD

**Issue by philipturner** [philipturner/metal-xla#2]
Referencing #1

**Comment by philipturner** [comment]
Right now, S4TF still isn't compiling with the new workflow. I need to build it as a Swift package, and get it to compile on Colab/Linux. In the meantime, you have a window to do the preliminary research stage of this effort as outlines in option 2 in the #1 thread.

**Comment by RahulBhalley** [comment]
Okay, got it

**Comment by philipturner** [comment]
That's part of it, but I also want to be able to find large segments of the stream that are compatible with MPSGraph, and run those on MPSGraph. Just using raw MPS like Torch is doing is an easier task, with lower performance. Also, don't use Objective-C like they do :-)

A Swift translation of some of Torch's code might be a good place to start off.

**Comment by RahulBhalley** [comment]
> I also want to be able to find large segments of the stream that are compatible with MPSGraph, and run those on MPSGraph.

(Just trying to understand your idea better.) I think you're saying is you want to run ops (defined by user in Swift front-end) on MPSGraph if they are available in it (say, `MPSGraph.multiplication`) but if some op is not available in MPSGraph (say, `stack`) then that should run on CPU (but how? using which framework). Do I get it right? If yes, wouldn't that already be defined in a given front-end function where the backend case will be handled with conditional statements?

**Comment by philipturner** [comment]
What you said about a conditional statement in the last sentence was a bit confusing. Conditional operators are fully supported operations in MPS Graph, as are some basic control flow constructs. Do you mean conditional as in it rejects an incompatible operator?

For CPU, we can rely on S4TF's CPU code for now (I think they have naive Swift code for every operator). There's different levels of fallback:

GPU mode:
MPSGraph: best
MPS: 2nd best
Accelerate: 3rd best
S4TF or naive custom Swift code: 4th best

CPU mode:
MLCompute: best
Accelerate: 2nd best
S4TF or naive custom Swift code: 3rd best

There are some interesting inconsistencies with S4TF and MPS. Random number generators are part of MPSGraph and even the XLA instruction set, but I think S4TF forces them onto the CPU with custom Swift code. I don't know whether we should put them on the GPU, because that brings behavior that doesn't 100% perfectly match the naive Swift code.

**Comment by philipturner** [comment]
Actually, it is possible to use MPSGraph as a run-session like API. That's why it's so compatible with GPU. I'll explain later, but it comes down to the ability to alias existing GPU memory and commit on a command queue.

This is also the reason I prefer MLCGraph for CPU, because you can access existing CPU memory from MLCTensors with 0 or 1 copies, and it's a commit-like API because the overhead of CPU function calls is almost nothing compared to GPU command encoding.

**Comment by philipturner** [comment]
The MPSGraph must be pre-compiled though, but that's exactly what happens when you compile training loops for XLA and then run the compiled graph multiple times. The first time (or multiple) will be spent executing eagerly (using raw MPS and not MPS Graph), but will switch to the more optimized MPSGraph executable once it finishes compiling.

**Comment by RahulBhalley** [comment]
We'll be using commit approach, right? as [`compile(with:feeds:targetTensors:targetOperations:compilationDescriptor:)`](https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/3787575-compile) seems like require the whole graph before executing it.

Interesting, you're right, looks like we can encode (and, hence, commit) for MPSGraph too [`encode(to:feeds:targetTensors:targetOperations:executionDescriptor:)`](https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/3580383-encode).

**Comment by philipturner** [comment]
Yes! That's what I saw that made me realize I could use MPSGraph for this!

**Comment by philipturner** [comment]
MLCompute's approach requires raw memory pointers, and Metal textures (even on shared memory) with layers can't be accessed as buffers without copying their memory. Thus, MLCompute won't work with raw MPS ops that must operate on textures. 

However, on the CPU, everything is done in on a buffer, so MLCGraph can work alongside external ops for that. Furthermore, MLCGraph is the only compiler that works for the CPU.

**Comment by RahulBhalley** [comment]
> Yes! That's what I saw that made me realize I could use MPSGraph for this!

Awesome, finally we're getting on same track.

**Comment by philipturner** [comment]
We will also need to delegate to MLCGraph is a tensor exceeds the maximum Metal texture size and MPSGraph crashes instead of changing it to a buffer-based op.

**Comment by RahulBhalley** [comment]
> However, on the CPU, everything is done in on a buffer, so MLCGraph can work alongside external ops for that. Furthermore, MLCGraph is the only compiler that works for the CPU.

Wow, it's like we'll be putting the best of MPS-derived frameworks into single package optimized for both CPU and GPU (and possibly also ANE for small set of ops). The result will be truly amazing!

**Comment by philipturner** [comment]
ANE won't work at all. It's only compatible with 8-bit integers and 16-bit half precision. Gradients can't propagate through either type. Only if Apple supported BFloat16 instead!

**Comment by RahulBhalley** [comment]
> ANE won't work at all. It's only compatible with 8-bit integers and 16-bit half precision. Gradients can't propagate through either type. Only if Apple supported BFloat16 instead!

Could be used for half precision inference in MetalXLA instead later.

**Comment by philipturner** [comment]
The initial research stage will be to experiment with this mess of interoperation and edge cases. I anticipate it's a lot of work that's just experimenting, and completely independent of S4TF (thus giving me a ~1 month window for getting S4TF to compile).

**Comment by philipturner** [comment]
> Could be used for half precision inference in MetalXLA instead later.

At that point, it isn't really work our effort to access MLCompute just for ANE. The pipeline of optimized training is vastly different from the pipeline of optimized inference. It would be cool, yes, but training is the priority of S4TF and we're running on a constrained time budget.

**Comment by RahulBhalley** [comment]
Okay, will try some experiments with MPS & MPSGraph operations using commit approach in a private repo adding you as collaborator.

**Comment by philipturner** [comment]
Why not make it public? That would set us apart from Torch who's currently doing it privately.

**Comment by philipturner** [comment]
All you need to do is not put anything silly when tearing up your code for debuggin. For DL4S-Experiments, I put a lot of "yeet" and other stuff, so I'd rather not make that public. Also, I just made all the commit messages "z" because that was the easiest to type on my keyboard.

For Swift-Colab, I had a much different experience and it's not too difficult to make the experimentation stage open-source.

**Comment by RahulBhalley** [comment]
> Why not make it public? That would set us apart from Torch who's currently doing it privately.

My code will be naive. 😅 Once I'm confident that we're both thinking the same thing (which will be reflected in code organization, ofc) then we can make it public and continue further experimentation.

> I put a lot of "yeet" and other stuff. 

Don't worry, I comment my code in a presentable form.

**Comment by philipturner** [comment]
> Once I'm confident that we're both thinking the same thing (which will be reflected in code organization, ofc) then we can make it public and continue further experimentation.

I'm okay with the experimentation stage being private. Just make it public as soon as we progress past that.

**Comment by RahulBhalley** [comment]
> You'll need to set up an organization to collaborate on a private repo.

Check your email 📧.

**Comment by philipturner** [comment]
Got it. I cancelled the xlakit invite.

**Comment by RahulBhalley** [comment]
Okay. I'll do some work then update the repo in a day or two.

**Comment by philipturner** [comment]
Let's keep our conversations as issues in this repository so they can be preserved for future reference. I don't know whether we'll ever make the private repo public, and it's harder to access because you have to swap profiles to access it.

**Comment by RahulBhalley** [comment]
No issues. We'll definitely make it public. Work is always better shown than kept private.

**Comment by RahulBhalley** [comment]
Also how should we approach the gradient computation? Using the available methods in MPS-derived frameworks?

**Comment by philipturner** [comment]
I doubt we can get that to work. I was just thinking about it, but it’s vastly more complex to track which set of operations is the gradient of which, especially with the define-by-run approach. Make 2 MPSGraphs, each with gradient pass disabled. One will be forward operators, the other backward.

**Comment by philipturner** [comment]
It looks like there is no naive Swift version for many essential operators (such as convolutions). They absolutely require interfacing with the same backend as Python TensorFlow. This is a big problem for iOS, which I highly doubt CTensorFlow will work on. Luckily, we have Accelerate to use and DL4S to refer to for those critical ops.

3D convolutions are a big problem, as neither DL4S or MPS has support for them. While you do the research on stitching together all the frameworks, I'll tabulate all the operators and which ones are concerning. This is solvable, we just need to manage time carefully to beat PyTorch.

**Comment by philipturner** [comment]
Its looking like I'll have to make an XLAKit framework to wrap MetalXLA and a possible DirectXLA. If I continue to discover evidence supporting this solution, I'll invite you to the organization again and make major changes to my plans. This backend would replace the entirety of CTensorFlow, not just the raw ops.

**Comment by RahulBhalley** [comment]
> This backend would replace the entirety of CTensorFlow, not just the raw ops.

Then the project will no longer be Swift _for TensorFlow_.

**Comment by philipturner** [comment]
No. It is S4TF because of the public API and the connection to XLA. I didn’t even know that it used TensorFlow as a backend until recently, and I don’t think most other developers care either.

**Comment by philipturner** [comment]
You don’t understand. Just about everybody can do inference. The problem is that we can’t do _training_ on Apple devices. It’s a very, very massive difference that most people ignore when they shouldn’t. Please don’t be the same way.

**Comment by philipturner** [comment]
When I said “I doubt we can get that to work”, I was referring to MPSGraph’s built in autograd. Due to conceptual complexity, we use Swift’s autodiff instead. There is no problem with training models, in fact that’s the only thing S4TF does.

**Comment by RahulBhalley** [comment]
Aah 😯 okay.

**Comment by philipturner** [comment]
I'm tabulating all of the ops MetalXLA needs to support and contemplating a name change for MetalXLA and XLAKit. "XLAKit" might be replaced by a revived SwiftRT that uses Apple/DirectX instead of CUDA, as SwiftRT is similar to the solution I'm thinking of. This is still very preliminary and I'm very unsure of what will happen with regard to it.

I just found out that MPSGraph has a hidden operator that allows for 3D convolutions through a little hack, so I might not need to collaborate with Torch. No longer needing to create "MetalConv3D" clears time for us to work on other ops. If we end up needing to work for several months before a final product (even after Torch's public beta), will you be okay with sticking to this effort until then?

**Comment by RahulBhalley** [comment]
> I just found out that MPSGraph has a hidden operator that allows for 3D convolutions through a little hack, so I might not need to collaborate with Torch. No longer needing to create "MetalConv3D" clears time for us to work on other ops.

Great 👍

> If we end up needing to work for several months before a final product (even after Torch's public beta), will you be okay with sticking to this effort until then?

Absolutely. Frankly, I think this work will take longer than LibTorch's stable release (I think they're already in beta stage i.e. shared the code on GitHub) as PyTorch has a larger team and more expertise (than me at least).

**Comment by RahulBhalley** [comment]
Btw that `encode` method works just like it does in MPS. I think we'll need a (possibly new) `Tensor` structure to store tensor data (of course) and also the backend information (as `enum`). From that backend information, we can run the op on a certain backend. Although this will be slow but will also allow us to move tensor between CPU and accelerators. This will address an issue when a certain op doesn't exists for a certain backend. 

_Disclaimer_: this is an idea but IDK how others do it currently.

**Comment by philipturner** [comment]
Your ideas seem just right. That’s just like how CTensorFlow does it, and how MPS works too (sort of). Your ideas are on the right track! I should finish tabulating all the ops and which are missing from Metal/DirectX today, which might be helpful to you.

**Comment by philipturner** [comment]
I just finished tabulating all of the ops and I have a solid plan. I may make some trivial Metal shaders to avoid switching from GPU to CPU when using eager execution (MPSGraph has several math functions that MPS doesn't have). I also learned some interesting stuff:
- MPSGraph is the most feature-complete of all Apple frameworks (the only one with 3D ops)
- MPSGraph is a superset of MPS except for some niche matrix operations
- MLCompute is the least feature-complete
- MLCompute is almost a subset of BNNS
- DirectML has every single operator needed (they're merging with TensorFlow right now)

Here's the publicly available link if you'd like to see it: [Tabulated S4TF Raw Ops](https://docs.google.com/spreadsheets/d/1d1TYUGOgutrv0g-decSoHih_-4VbXJn8f6J4aLaa4Fc/edit?usp=sharing)

**Comment by RahulBhalley** [comment]
> Here's the publicly available link if you'd like to see it: [Tabulated S4TF Raw Ops](https://docs.google.com/spreadsheets/d/1d1TYUGOgutrv0g-decSoHih_-4VbXJn8f6J4aLaa4Fc/edit?usp=sharing)

So the XLA ops marked **x** are available or do they fall back to CPU?

**Comment by philipturner** [comment]
That was just data I gathered to cull the XLA-exclusive ops I didn't need to include in the Metal backend's instruction set. They're mostly irrelevant now; the rightmost columns are more helpful for finding what falls back to the CPU.

**Comment by RahulBhalley** [comment]
Found something, but you might already be aware of it. 

The XLA ops registration for [PyTorch](https://github.com/pytorch/xla/tree/master/torch_xla/csrc/ops) and [S4TF](https://github.com/tensorflow/swift-apis/tree/main/Sources/x10/xla_tensor/ops) are highly similar. And because PyTorch is active, it has many more ops available which we can use for S4TF.

**Comment by philipturner** [comment]
I'm sorry, but the similarities actually don't help. Both are just wrappers over CTensorFlow. However, when Torch releases their Metal solution, we can "borrow" any solutions they come up with for GPU versions of rare ops (although not on Windows). If you want to look for solutions to some of the problematic ops in the spreadsheet right now instead of drafting interoperation between Apple ML frameworks, that's fine as well. As long as you're exploring, any contribution is helpful.

Keep in mind that I'm also targeting Windows with this backend (just to make it possible, not 100% sure it will happen). Any additional ops we add from Aten to improve performance (as pre-compilation operator fusion) might also need to run on DirectML or at least have the Torch CPU source code copied to our backend. I think this backend may need to replace slightly more than CTensorFlow, enough to modify behavior at `_Raw` binding points to solve a few problems in the spreadsheet.

Also, we might be able to use Swift autodiff to get a mega slow version of Cholesky grad and bilinear/lanczos resample grad that at least doesn't cause a fatal error. Those ops really don't belong in ML, but they're already ingrained in S4TF's public API.

**Comment by philipturner** [comment]
CTensorFlow already covers CUDA and TPU.

**Comment by philipturner** [comment]
We aren't forking any existing stuff, we're keeping the same idea as before but just making a name more likely to be accepted by TensorFlow and merged into the main branch.

**Comment by philipturner** [comment]
PlaceholderDevice sounds like a wonderful name - I'll think about it.

**Comment by RahulBhalley** [comment]
Because we will use Swift's differentiation feature, I think we have to discard the idea of using MPSGraph's differentiation system as it doesn't seems possible to merge these two. This will also alleviate the problem of tracking ops graph as it will already be recorded by Swift's differentiation. Correct me if I'm wrong. 

> PlaceholderDevice sounds like a wonderful name - I'll think about it.

Sure.

**Comment by philipturner** [comment]
I already talked about not using MPSGraph's autodiff in a prior comment. It was the one that caused a big misunderstanding and ending with you reacting with an emoji.

**Comment by RahulBhalley** [comment]
Wouldn't it be better if we expose `MPSGraphTensorData` to `Tensor` of S4TF via a C wrapper like CTensorFlow (I think I read somewhere you telling me we don't need Objective C access MPS but C), if possible? This will potentially solve the problem I mentioned in my Swift Forum's [post](https://forums.swift.org/t/unable-to-define-custom-derivative-for-mpsgraph-methods/54370) also other MPS-derived framework's tensor storage constructs could be made differentiable.

**Comment by philipturner** [comment]
No... MPSGraphTensorData is just an abstract construct that doesn't even materialize unless you explicitly request it during graph compilation. A C wrapper through Core Foundation won't change anything. Plus, CTensorFlow doesn't run iOS, so we can't use that.

**Comment by philipturner** [comment]
What I said about it not materializing unless you request so isn't entirely the problem, it's just that every possible backend communicates with S4TF through the same interface. We could an MPSGraphTensor to back the underlying storage of all of our tensors, although I don't know for sure if we can pull that off for MLCTensor. We also need to work with discrete GPU memory and Metal textures whose data can't be accessed through a pointer without copying.

**Comment by philipturner** [comment]
Have you gotten around to experimenting with anything yet?

**Comment by philipturner** [comment]
From the lack of activity, it seems like you have something else going on. Am I correct? There's no pressure for you to do anything, I'm happy that you're even interested in talking about this.

**Comment by RahulBhalley** [comment]
Hi Philip, sorry for delayed reply. You are right, it's just hard to devote time to any work more than just a full-time job.
